chatsky/core/ctx_dict.py

```diff
@@ -0,0 +1,476 @@
+        Also hashes will be calculated for the initially loaded items for modification tracking.
+
+        :param storage: Context storage, keeping the current context.
+        :param id: Newly created context ID.
+        :param field: Current dict field name.
+        :return: New "connected" context dict.
+        """
+
+        logger.debug(f"Connected context dict created for {id}, {field}")
+        keys, items = await gather(storage.load_field_keys(id, field), storage.load_field_latest(id, field))
+        val_key_items = [(k, v) for k, v in items if v is not None]
+        logger.debug(f"Context dict for {id}, {field} loaded: {collapse_num_list(keys)}")
+        instance = cls()
+        instance._storage = storage
+        instance._ctx_id = id
+        instance._field_name = field
+        instance._keys = set(keys)
+        instance._items = {k: instance._value_type.validate_json(v) for k, v in val_key_items}
+        instance._hashes = {k: _get_hash(v) for k, v in val_key_items} if storage.rewrite_existing else dict()
+        return instance
+
+    async def _load_items(self, keys: List[int]) -> None:
+        """
+        Load items for the given keys from the connected context storage.
+        Update the `_items` and `_hashes` fields if necessary.
+        NB! If not all the requested items are available,
+        only the successfully loaded will be updated and no error will be raised.
+
+        :param keys: The requested key array.
+        """
+
+        logger.debug(
+            f"Context dict for {self._ctx_id}, {self._field_name} loading extra items: {collapse_num_list(keys)}..."
+        )
+        items = await self._storage.load_field_items(self._ctx_id, self._field_name, keys)
+        logger.debug(
+            f"Context dict for {self._ctx_id}, {self._field_name} extra items loaded: {collapse_num_list(keys)}"
+        )
+        for key, value in items:
+            self._items[key] = self._value_type.validate_json(value)
+            if not self._storage.rewrite_existing:
+                self._hashes[key] = _get_hash(value)
+
+    @overload
+    async def __getitem__(self, key: int) -> BaseModel: ...  # noqa: E704
+
+    @overload
+    async def __getitem__(self, key: slice) -> List[BaseModel]: ...  # noqa: E704
+
+    async def __getitem__(self, key):
+        if isinstance(key, int) and key < 0:
+            key = self.keys()[key]
+        if self._storage is not None:
+            if isinstance(key, slice):
+                await self._load_items(
+                    [self.keys()[k] for k in range(len(self.keys()))[key] if k not in self._items.keys()]
+                )
+            elif key not in self._items.keys():
+                await self._load_items([key])
+        if isinstance(key, slice):
+            return [self._items[k] for k in self.keys()[key]]
+        else:
+            return self._items[key]
```

Comment:
What I don't like about the way slices are processed now is that single key refers to a specific turn while a slice refers to a range of available turns.
So if there are only turns 1, 3, 5 available in ctx_dict, `ctx_dict[1:2] == ctx_dict[3], ctx_dict[5]`, which is very confusing.

I think that there's little use case to getting ctx_dict items by their position in the keys.
So my proposal is:
Allow only
1. `ctx_dict[non-neg-int]` -- return element on turn `non-neg-int` (turn number is a non-negative integer);
2. `ctx_dict[non-neg-int1:non-neg-int2:non-neg-int3]` -- return elements starting from turn `non-neg-int1` up to `non-neg-int2` with a step of `non-neg-int3`.

(just in case: this proposal is for all the methods)